### PR TITLE
Fix/push literal variable

### PR DESCRIPTION
**Pull request by guillep**
Depends on https://github.com/Alamvic/druid/pull/48 and https://github.com/pharo-project/pharo-vm/pull/586
Add support for push Literal Variable

This includes the main change that bytecodes should have a continuation for the next bytecode so everything compiles fine regardless the order in which basic blocks are evaluated.

### CODE REVIEW

File: Druid/DRBytecodeIRGenerator.class.st [hunk 1]
```diff
@@ -289,6 +289,20 @@ DRBytecodeIRGenerator >> push: aValue [
...
+	| frame |
+	frame := super pushFrameForCode: aDRCode receiver: aReceiver.
+
+	"If this is the base frame, it should continue with the next bytecode at the end"
+	executionState baseFrame = frame ifTrue: [
+		frame exitBasicBlock endInstruction:
+			(self instantiateNoResultInstruction: DRContinueNextBytecode) ].
+
+	^ frame
+]
```
Comment (PalumboN) [inline Druid/DRBytecodeIRGenerator.class.st:304]: Also exists the method `finishCodeInterpretation: lastFrame` that is called when the last frame is popped (I don't know if it's the same to add the instruction on push or pop)

File: Druid/DRBasicBlock.class.st [hunk 2]
```diff
@@ -345,8 +345,7 @@ DRBasicBlock >> greater: operand1 than: operand2 [
 { #category : #testing }
 DRBasicBlock >> hasFinalInstruction [
 
-	^ (endInstruction notNil and: [ endInstruction isNoop not ]) or: [ 
-		  instructions anySatisfy: [ :i | i isFrameReturn ] ]
```
Comment (PalumboN) [inline Druid/DRBasicBlock.class.st:349]: This doesn't break the inlining!? Woow

File: Druid/DRPhiFunction.class.st [hunk 3]
```diff
@@ -205,9 +205,24 @@ DRPhiFunction >> simpleConstantFold [
...
+
+	[ worklist isEmpty ] whileFalse: [
+		| next |
+		next := worklist removeFirst.
+		(seen includes: next) ifFalse: [
+			next isPhiFunction
+				ifTrue: [ 
+					seen add: next.
+					worklist addAll: next operands ]
+				ifFalse: [ simplifiedOperands add: next ] ] ].
```
Comment (PalumboN) [inline Druid/DRPhiFunction.class.st:225]: You can use `whileNotEmpty:` for this 🤔 

```suggestion
	(OrderedCollection withAll: operands) whileNotEmpty: [:next :col :marked | 
		(marked includes: next) ifFalse: [
			next isPhiFunction
				ifTrue: [ 
					marked add: next.
					col addAll: next operands ]
				ifFalse: [ simplifiedOperands add: next ] ]  ].
```

File: Druid/DRCogitStackToRegisterMappingGenerator.class.st [hunk 4]
```diff
@@ -53,6 +53,20 @@ DRCogitStackToRegisterMappingGenerator >> visitCogitSendMarshall: aDRCogitSendMa
...
+	
+	"This is a continuation to the next bytecode.
+	If this is the last basic block, do nothing.
+	Otherwise, make a jump to the end"
+	self isLastBasicBlock ifTrue: [ ^ self ].
+	
+	"Jump to the end!"
+	self
+		generateJump: #Jump:
+		target: nil
```
Comment (PalumboN) [inline Druid/DRCogitStackToRegisterMappingGenerator.class.st:67]: Can we use as target `#nextBytecode`?

**Review by PalumboN — COMMENTED** [review]
Cool! I left some comments. The PR can be merged when all tests are fixed (now there are some MDU)

File: Druid/DRBytecodeIRGenerator.class.st (hunk 1 above)
Comment (guillep) [reply Druid/DRBytecodeIRGenerator.class.st:304]: Yeah, I don't know if adding it at the end or at the beginning of the frame...

File: Druid/DRBasicBlock.class.st (hunk 2 above)
Comment (guillep) [reply Druid/DRBasicBlock.class.st:349]: No idea. I broke a couple of tests, do not seem closely related to inlining


DRBytecodeGeneratorTest
 ✗ #testMethodRecompilationWithProfile (82ms)

DRLinearizationTest
 ✗ #testBuildCFGHasAllBlocks (69ms)
 ✗ #testBuildCFGHasAllBlocksWhenEarlyMethodReturns (54ms)
 ✗ #testBuildCFGHasAllBlocksWithTwoNonLocalReturns (59ms)
 ✗ #testBuildCFGHasAllBlocks (63ms)
 ✗ #testBuildCFGHasAllBlocksWhenEarlyMethodReturns (53ms)
 ✗ #testBuildCFGHasAllBlocksWithTwoNonLocalReturns (53ms)

DRCogitCanonicaliserTest
 ✗ #testSimplifyAdditionOfAddition (2ms)
 ✗ #testSimplifyLoadAdditionOfAddition (1ms)

File: Druid/DRCogitStackToRegisterMappingGenerator.class.st (hunk 4 above)
Comment (guillep) [reply Druid/DRCogitStackToRegisterMappingGenerator.class.st:67]: yes

File: Druid/DRCogitCodeGenerator.class.st [hunk 5]
```diff
@@ -111,6 +135,8 @@ DRCogitCodeGenerator >> generateCodeForCFG: aDRControlFlowGraph [
 		fallthroughBlock := blocks at: i + 1 ifAbsent: nil.
 		self preProcessBlock: b.
 		b instructionsDo: [ :e | self buildStatementsForInstruction: e ] ].
+	"Generate the continuation jumps to the following code"
+	self generateBlockIncomingJumps: nil.
```
[the remark this answers is not in the document]
Comment (guillep) [reply Druid/DRCogitCodeGenerator.class.st:139]: Yes

File: Druid/DRBasicBlock.class.st (hunk 2 above)
Comment (guillep) [reply Druid/DRBasicBlock.class.st:349]: So, testMethodRecompilationWithProfile is probably related, I tried to fixed it but it is too tangled on how the bytecode is generated...

**Comment by guillep** [comment]
Tests are green. Let's discuss the skipped test tomorrow.

File: Druid/DRBasicBlock.class.st (hunk 2 above)
Comment (PalumboN) [reply Druid/DRBasicBlock.class.st:349]: Ahhh that one can be skipped, it was the last experiment with DrOpal so everything is very fragile. It's ok!

**Pull request merged**